README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# <span style="color: #2395ec;">dev</span><span style="color: #38424f;">in</span><span style="color: #2395ec;">docker</span> <img src="man/figures/logo.png" align="right" alt="" width="120" />

<!-- badges: start -->

[![R build
status](https://github.com/ThinkR-open/devindocker/workflows/R-CMD-check/badge.svg)](https://github.com/ThinkR-open/devindocker/actions)
<!-- badges: end -->

The goal of {devindocker}, as “Development In Docker”, is to help launch
your R project inside a Docker container with Rstudio server.

  - You can develop with the same architecture as your clients
  - You can test your project / package on a specific architecture
  - You can combine this with {renv} to keep packages versions installed
    inside your directory (and thus available next time you start the
    container)
  - All changes in settings of your Rstudio Server will be kept for next
    start (last project opened, editor theme, …)

## Installation

You can install the development version from
[GitHub](https://github.com/) with:

``` r
# install.packages("remotes")
remotes::install_github("ThinkR-open/devindocker")
# Or with vignettes included
remotes::install_github("ThinkR-open/devindocker", build_vignettes = TRUE)
```

## Usage

Let us create a random directory with a file inside.

``` r
# Temporary project
tempdir <- tempdir()
path <- file.path(tempdir, "myproject")
dir.create(path)
# Add a file inside
cat("# my R file", file = file.path(path, "my-file.R"))
```

### Work in a Docker environment

Launch a Docker container with your directory inside. This should be a
container with Rstudio server inside.  
*Note that you start outside your project, which means you will have to
start a new RStudio project if this is your way of working.*  
*Note that packages you install will not be kept after you stop the
container, but RStudio preferences will.*

``` r
library(devindocker)
# Which path to your working directory / project
path <- file.path(tempdir, "myproject")

# Which container (with Rstudio inside) ? ----
# https://hub.docker.com/r/rocker/verse
container <- "rocker/geospatial:4.0.1"

# Which port ? ----
# _Useful if multiple Rstudio Server to launch
port <- 8788

# Start Docker project ----
launch_proj_docker(
  path = path,
  container = container,
  port = port)
```

When you’re done, do not forget to stop properly the Rstudio Server:
Click on Top right button to quit or `q()` in the console.

Then, end the container.

``` r
# Stop Docker properly
stop_proj_docker(path = path)
```

### Use {renv} inside Docker and keep installation of packages

**Note that you need to launch your project with {devindocker} from
outside this project. Never ever open it again locally (outside a Docker
container) if you want to avoid problems with bad and not compatible
local {renv} setup. It is recommended to create a project dedicated to
launch {devindocker} projects.**

Launch a Docker container with your directory inside. This should be a
container with RStudio Server inside.  
*Note that you start outside your project, which means you will have to
start a new RStudio project if this is your way of working.*  
*Note that packages you install will be kept after you stop the
container, as well as RStudio preferences.*

**Follow instructions in the `"renv_instructions.Rmd"` file that is
created inside your project.**

``` r
# Which path to your working directory / project
path <- file.path(tempdir, "myproject")

# Which container (with Rstudio inside) ? ----
# https://hub.docker.com/r/rocker/verse
container <- "rocker/geospatial:4.0.1"

# Which port ? ----
# _Useful if multiple Rstudio Server to launch
port <- 8788

# My renv cache directory on my local computer
# Used as persistent drive for all you Docker container with {devindocker}
renv_cache <- "~/renv_cache"

# Start Docker project ----
devindocker::launch_proj_docker(
  path = path,
  container = container,
  port = port,
  renv_cache = renv_cache,
  renv_inst = TRUE, # Add an Rmd with instructions inside your project
  update_docker = TRUE
)
```

When you’re done, do not forget to stop properly the Rstudio Server:
Click on Top right button to quit or `q()` in the console.

Then, end the container.

``` r
# Stop Docker properly
stop_proj_docker(path = path)
```

### Connect a container to a Docker network

You can connect your container to a Docker network using parameter
`network_name`.  
See [vignette for an example wih mysql
here](https://thinkr-open.github.io/devindocker/articles/ac-docker-network.html).

## Install Docker on your OS

Here is the list of resources written by Docker to install Docker on
your computer depending on your OS:

  - Windows 10 64-bit: Pro, Enterprise, or Education (Build 16299 or
    later) : <https://docs.docker.com/docker-for-windows/install/>
  - Install Windows 10 Home, version 1903 or higher. :
    <https://docs.docker.com/docker-for-windows/install-windows-home/>
  - MacOS : <https://docs.docker.com/docker-for-mac/install/>
  - Linux: <https://docs.docker.com/engine/install/#server>
      - Ubuntu: <https://docs.docker.com/engine/install/ubuntu/>

## Code of Conduct

Please note that the devindocker project is released with a [Contributor
Code of
Conduct](https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html).
By contributing to this project, you agree to abide by its terms.


NEWS.md
# devindocker 0.0.1.9000

## Breaking 

* Parameter `project_path` is now named `path` as any directory path can be used.
Not only projects.

## Major

* Allow to add additional volumes using parameter `volumes`. See Vignette.
* Allow to change default url using parameter `url`
* Allow connection to a Docker network with parameter `network_name`. See Vignette.
* Allow to start container with root privileges with parameter `root = TRUE`. 

## Minor

* Allow to store library out of the project when using {renv}. See Vignette.

# devindocker 0.0.1

* Added a `NEWS.md` file to track changes to the package.
* First release


tests/testthat.R
library(testthat)
library(devindocker)

test_check("devindocker")


tests/testthat/test-launch_proj_docker.R
# Create a package directory with files
tempdir <- tempdir()
my_project <- normalizePath(file.path(tempdir, "myproject"), mustWork = FALSE)
usethis::create_package(my_project, open = FALSE)

# Which container with RStudio server? ----
container <- "rocker/geospatial:4.0.1"
# Which port ? ----
port <- Sys.getenv("PORT", unset = 8787)
url <- Sys.getenv("URL", unset = "http://127.0.0.1")

test_that("launch_proj_docker and stop work", {
  # skip_on_ci()
  # skip_on_cran()

  # Start Docker project
  output <- launch_proj_docker(path = my_project,
                     container = container,
                     port = port,
                     open_url = FALSE)

  # RStudio server has started
  get_html <- httr::GET(url = paste0(url, ":", port))
  expect_equal(get_html$status_code, 200)
  # RStudio config files exist
  expect_true(dir.exists(file.path(my_project, "rstudio_dotrstudio")))
  expect_true(dir.exists(file.path(my_project, "rstudio_dotconfig")))

  # Stop Docker properly
  stop_proj_docker(path = my_project, sleep = 5)
  # RStudio server has stopped
  expect_error(httr::GET(url = paste0(url, ":", port)))
  # RStudio config files still exist
  expect_true(dir.exists(file.path(my_project, "rstudio_dotrstudio")))
  expect_true(dir.exists(file.path(my_project, "rstudio_dotconfig")))
})

# Test volumes ----
my_additional <- normalizePath(file.path(tempdir, "my_additional"), mustWork = FALSE)
dir.create(my_additional)
my_additional2 <- normalizePath(file.path(tempdir, "my_additional2"), mustWork = FALSE)
dir.create(my_additional2)

## Create the table of correspondance for additional volumes
volumes <- data.frame(
  local = c(my_additional, my_additional2),
  docker = c("/home/rstudio/my_additional", "/home/rstudio/my_additional2")
)

test_that("other volumes work", {
  # Start Docker project
  output <- launch_proj_docker(path = my_project,
                               container = container,
                               port = port,
                               volumes = volumes,
                               open_url = FALSE)

  # RStudio server has started
  get_html <- httr::GET(url = paste0(url, ":", port))
  # expect_equal(get_html$status_code, 200)

  # Add a file in the additional directory from inside the container
  system(
    paste0("docker exec ", devindocker:::clean_project_name(basename(my_project)),
           " Rscript -e 'cat(\" test\", file = \"", volumes[1, "docker"], "/test.txt\")'")
  )
  expect_true(file.exists(file.path(volumes[1,"local"], "test.txt")))

  system(
    paste0("docker exec ", devindocker:::clean_project_name(basename(my_project)),
           " Rscript -e 'cat(\" test\", file = \"", volumes[2, "docker"], "/test2.txt\")'")
  )
  expect_true(file.exists(file.path(volumes[2,"local"], "test2.txt")))

  # Stop Docker properly
  stop_proj_docker(path = my_project, sleep = 5)
  # RStudio server has stopped
  expect_error(httr::GET(url = paste0(url, ":", port)))
})
unlink(my_project, recursive = TRUE)
unlink(volumes[1,"local"], recursive = TRUE)
unlink(volumes[2,"local"], recursive = TRUE)


# _Not recommended dir name ----
my_bad_project <- normalizePath(file.path(tempdir, "my Project"), mustWork = FALSE)
dir.create(my_bad_project)

test_that("not recommended dir name work", {
  # Start Docker project
  output <- launch_proj_docker(path = my_bad_project,
                               container = container,
                               port = port,
                               open_url = FALSE)

  # RStudio server has started
  get_html <- httr::GET(url = paste0(url, ":", port))
  # expect_equal(get_html$status_code, 200)

  # Stop Docker properly
  stop_proj_docker(path = my_project, sleep = 5)
  # RStudio server has stopped
  expect_error(httr::GET(url = paste0(url, ":", port)))
})
unlink(my_bad_project, recursive = TRUE)


dev_history.R
usethis::create_package("devindocker")
usethis::use_build_ignore("dev_history.R")
usethis::use_build_ignore("_pkgdown.yml")

# License
usethis::use_mit_license("ThinkR")
usethis::use_code_of_conduct()

# Doc
usethis::use_readme_rmd()
usethis::use_vignette("aa-dev-in-docker")
usethis::use_vignette("ab-dev-in-docker-with-renv")
usethis::use_vignette("ac-docker-network")
usethis::use_vignette("ac-docker-volumes")
usethis::use_github_action_check_standard()
usethis::use_github_action("pkgdown")
pkgdown::build_site()
usethis::use_news_md()
usethis::use_package_doc()

# Tests
usethis::use_testthat()
usethis::use_test("launch_proj_docker")

# devindocker inside devindocker
# _Start Docker project
origwd <- setwd(dirname(normalizePath(".")))
container <- "rocker/geospatial:4.0.1"
launch_proj_docker(path = "devindocker",
                   container = container,
                   port = 8080)
# _Stop Docker properly
stop_proj_docker(path = "devindocker", sleep = 2)
setwd(origwd)


# Dev
attachment::att_amend_desc()

options(rmarkdown.html_vignette.check_title = FALSE)
devtools::check()



R/launch-stop-docker.R
#' Launch and stop Docker
#'
#' @param path Path to project or any directory to launch.
#' @param container Docker container to download from docker hub.
#' @param network_name Character. Give the name of the network in which the
#'  container will be included using \code{--net network_name}.
#' @param port Local port to which to launch Rstudio Server
#' @param renv_inst Logical. Whether to add a R script with {renv} instructions in the project.
#' @param renv_cache Path to renv cache on your computer. Set to FALSE to not use renv.
#' @param renv_out Whether to set {renv} libraries out of the project.
#' @param renv_out_dir Where to store project libraries. Default to a ".renv" folder
#' in the parent directory of "path"
#' @param update_docker Logical. Whether to update Docker container with DockerHub.
#' @param is_root logical. Whether the Docker user has root permissions (add to sudoers).
#' Can be useful if you want to simulate your CI behaviour in the Terminal using "\code{sudo R}".
#' @param volumes data.frame with two columns named `local` and `docker`.
#'  `local` contains path to local computer.
#'  `docker` contains path inside the docker container linked to local.
#' @param open_url Logical. Whether to open container URL in browser on http://127.0.0.1:port
#' @param url Localhost url. Default to "http://127.0.0.1". "localhost" can be a good alternative.
#'
#' @importFrom utils browseURL
#'
#' @details
#' renv_out : It is recommended to set {renv library} out of the project to avoid copy of entire library
#' during package checks. See \code{vignette("packages", package = "renv")}.
#'
#' @export
#' @examples
#' \dontrun{
#' tempdir <- tempdir()
#' path <- file.path(tempdir, "myproject")
#' usethis::create_package(path, open = FALSE)
#'
#' # Which Rstudio container ? ----
#' container <- c("thinkr/rstudio3_5_2",
#'                "rocker/geospatial:4.0.1")[2]
#'
#' # Which port ? ----
#' # _Useful if multiple Rstudio Server to launch
#' port <- 8788
#'
#' # _Start Docker project
#' launch_proj_docker(path = path,
#'                    container = container,
#'                    port = port)
#'
#' # _Stop Docker properly
#' stop_proj_docker(path = path, sleep = 5)
#'
#' # With renv cache shared with host
#' dir.create(file.path(tempdir, "cache"))
#' tempcache <- file.path(tempdir, "cache")
#' # _Start Docker project
#' launch_proj_docker(path = path,
#'                    container = container,
#'                    port = port,
#'                    renv_cache = tempcache)
#'
#' # _Stop Docker properly
#' stop_proj_docker(path = path, sleep = 5)
#'
#' # Mount with additional volumes
#' }

launch_proj_docker <- function(path = ".",
                               container = "thinkr/rstudio3_6_1_geo",
                               network_name = "r-db",
                               port = 8787,
                               renv_inst = FALSE,
                               renv_cache = FALSE,
                               renv_out = FALSE,
                               renv_out_dir,
                               update_docker = TRUE,
                               is_root = FALSE,
                               volumes,
                               open_url = TRUE,
                               url = "http://127.0.0.1"
                               # vbox = FALSE
) {
  # @param vbox Logical. If Docker run on windows in a virtual box, paths need to be changed

  path <- normalizePath(path)

  if (missing(renv_cache) | renv_cache == FALSE) {
    renv_cache <- NULL
  }

  # addtional volumes
  if (!missing(volumes)) {
    if (!all(c("local", "docker") %in% names(volumes))) {
      stop("To mount external volumes, use a data.frame with names: 'local' and 'docker'")
    }
    volumes[,"local"] <- normalizePath(volumes[,"local"])

    add_volumes <- paste(
      apply(volumes, 1, function(x) paste0(" -v '", x["local"], ":", x["docker"], "'")),
      collapse = ""
    )
  } else {
    add_volumes <- NULL
  }

  # First time ----
  if (!dir.exists(file.path(path, "rstudio_dotrstudio"))) {
    # Hide this file
    # rstudio last files and others
    dir.create(normalizePath(file.path(path, "rstudio_dotrstudio"), mustWork = FALSE))
  }
  if (!dir.exists(file.path(path, "rstudio_dotconfig"))) {
    # Hide this file
    # rstudio appearance preferences
    dir.create(normalizePath(file.path(path, "rstudio_dotconfig"), mustWork = FALSE))
  }
  if (isTRUE(renv_out)) {
    if (missing(renv_out_dir)) {
      renv_out_dir <- normalizePath(file.path(dirname(path), ".renv"), mustWork = FALSE)
    }
    if (!dir.exists(renv_out_dir)) {dir.create(renv_out_dir)}
  }

  # .gitignore
  # Files to ignore
  lines <- stats::na.omit(
    c("rstudio_dotrstudio", "rstudio_dotconfig",
      ifelse(renv_inst, "renv_instructions.Rmd", NA)
    )
  )

  gitfile <- normalizePath(file.path(path, ".gitignore"), mustWork = FALSE)
  if (!file.exists(gitfile)) {
    existing_lines <- ""
  } else {
    existing_lines <- readLines(gitfile, warn = FALSE, encoding = "UTF-8")
  }
  new <- setdiff(lines, existing_lines)
  if (length(new) != 0) {
    all <- c(existing_lines, new)
    cat(all, file = gitfile, sep = "\n")
  }

  # .Rbuildignore
  # Files to ignore
  lines <- stats::na.omit(
    c("rstudio\\_dotconfig",
      "rstudio\\_dotrstudio",
      ifelse(renv_inst, "renv\\_instructions\\.Rmd", NA)
    ))

  buildfile <- normalizePath(file.path(path, ".Rbuildignore"), mustWork = FALSE)
  if (!file.exists(buildfile)) {
    existing_lines <- ""
  } else {
    existing_lines <- readLines(buildfile, warn = FALSE, encoding = "UTF-8")
  }
  new <- setdiff(lines, existing_lines)
  if (length(new) != 0) {
    all <- c(existing_lines, new)
    cat(all, file = buildfile, sep = "\n")
  }

  # Pull container if needed
  if (isTRUE(update_docker)) {
    system(paste("docker pull", container))
  }

  # Future ----
  ## Allow us to lanch the system command in a new R session
  # library(future)
  # Requires at least 2 workers otherwise does not work
  future::plan(future::multisession(workers = 2))

  ## Launch ----
  dockername <- clean_project_name(basename(path))
  projectname <- clean_dir_name(basename(path))

  if (!is.null(renv_cache)) {
    # {renv} path in container
    RENV_PATHS_CACHE_CONTAINER <- "/opt/local/renv/cache"
    # RENV_PATHS_CACHE_CONTAINER <- "/home/rstudio/.local/share/renv/cache"
    RENV_PATHS_LIBRARY_ROOT_CONTAINER = "/home/rstudio/.renv/library"
    # Directory with all {renv} package cache on host
    RENV_PATHS_CACHE_HOST <- normalizePath(renv_cache, mustWork = FALSE)
    if (!dir.exists(RENV_PATHS_CACHE_HOST)) {
      dir.create(RENV_PATHS_CACHE_HOST)
      message(RENV_PATHS_CACHE_HOST, " was created")
    }
    # Directory with local copy of packages for the project
    if (isTRUE(renv_out)) {
      RENV_PATHS_LIBRARY_ROOT_HOST = renv_out_dir
    }
  }

  # Problem windows with virtualbox
  # if (.Platform$OS.type == "windows" & isTRUE(vbox)) {
  #   path <- normalizePath(path, winslash = "/", mustWork = FALSE)
  #   if (grepl("^[A-Z]:", path)) {
  #   path <- gsub("^([A-Z]):", "//\\1", path)
  #   }
  # }

  ## Create docker network
  try(system(paste("docker network create", network_name)))

  ## Launch the server in the new R session (terminal have to be active...)
  future::future({
    system(
      paste0(
        "docker run --name ", dockername,
        # network
        # ifelse(isTRUE(with_mysql), " --net r-db", ""),
        " --net ", network_name,
        # root permission
        ifelse(isTRUE(is_root), " -e ROOT=TRUE", ""),
        " -d -e DISABLE_AUTH=true",
        # {renv}
        # _Global renv cache
        ifelse(!is.null(renv_cache), paste0(" -e RENV_PATHS_CACHE=", RENV_PATHS_CACHE_CONTAINER), ""),
        ifelse(!is.null(renv_cache), paste0(" -v ", RENV_PATHS_CACHE_HOST, ":", RENV_PATHS_CACHE_CONTAINER), ""),
        # _Project renv library
        ifelse(isTRUE(renv_out), paste0(" -e RENV_PATHS_LIBRARY_ROOT=", RENV_PATHS_LIBRARY_ROOT_CONTAINER), ""),
        ifelse(isTRUE(renv_out), paste0(" -v ", RENV_PATHS_LIBRARY_ROOT_HOST, ":", RENV_PATHS_LIBRARY_ROOT_CONTAINER), ""),

        # Rstudio server
        " -v '", path, ":/home/rstudio/", projectname, "'",
        " -v '", normalizePath(file.path(path, "rstudio_dotconfig"), mustWork = TRUE), ":/home/rstudio/.config'", #/rstudio
        " -v '", normalizePath(file.path(path, "rstudio_dotrstudio"), mustWork = TRUE), ":/home/rstudio/.rstudio'",

        # Addtional volumes
        ifelse(!is.null(add_volumes), add_volumes, ""),

        " -p 127.0.0.1:", port, ":8787 ",
        container),
      intern = TRUE)
  })

  if (isTRUE(renv_inst)) {
    if (!file.exists(file.path(path, "renv_instructions.Rmd"))) {
      update_renv_help(path, overwrite = FALSE)
    }
  }
  Sys.sleep(5)
  if (isTRUE(open_url)) {
    browseURL(paste0(url, ":", port))
  }
}

#' Update renv instructions with last version
#'
#' @param path Path where to save the renv_instructions.Rmd file
#' @param overwrite Whether to overwrite existing file
#'
#' @export
update_renv_help <- function(path = "", overwrite = TRUE) {
  file.copy(
    system.file("renv", "renv_instructions.Rmd", package = "devindocker"),
    file.path(path, "renv_instructions.Rmd"),
    overwrite = overwrite
  )
}

#' Stop running Docker container
#'
#' @param sleep Numeric. Number of seconds to wait for user to correctly stop Rstudio Server
#' @param stop_network Logical. Whether to stop Docker network.
#'
#' @export
#' @rdname launch_proj_docker

stop_proj_docker <- function(path, sleep = 10, network_name = "r-db", stop_network = TRUE) {

  dockername <- clean_project_name(basename(path))

  message("
  # --- /!\ Do not forget to stop properly the Rstudio Server /!\ --- #
  # Click on Top right button to quit or `q()` in the console
  ")

  Sys.sleep(sleep)

  system(paste("docker kill", dockername))
  system(paste("docker rm", dockername))

  if (isTRUE(stop_network)) {
    try(system(paste("docker network remove", network_name)))
  }
}


clean_project_name <- function(x) {
  tolower(gsub(" ", "", x))
}
clean_dir_name <- function(x) {
  gsub(" ", "\\", x)
}
